integration-tests/src/tests/client/features/delegate_action.rs

```diff
@@ -572,3 +577,148 @@ fn assert_ft_balance(
     let balance = std::str::from_utf8(&response.result).expect("invalid UTF8");
     assert_eq!(format!("\"{expected_balance}\""), balance);
 }
+
+/// Test account creation scenarios with meta transactions.
+///
+/// Named accounts aren't the primary use case for meta transactions but still
+/// worth a test case.
+#[test]
+fn meta_tx_create_named_account() {
+    let relayer = bob_account();
+    let sender = alice_account();
+    let new_account = eve_dot_alice_account();
+    let node = RuntimeNode::new(&relayer);
+
+    let fee_helper = fee_helper(&node);
+    let amount = NEAR_BASE;
+
+    let public_key = PublicKey::from_seed(KeyType::ED25519, &new_account);
+
+    // That's the minium to create a (useful) account.
+    let actions = vec![
+        Action::CreateAccount(CreateAccountAction {}),
+        Action::Transfer(TransferAction { deposit: amount }),
+        Action::AddKey(AddKeyAction { public_key, access_key: AccessKey::full_access() }),
+    ];
+
+    // Check account doesn't exist, yet
+    node.view_account(&new_account).expect_err("account already exists");
+
+    let tx_cost = fee_helper.create_account_transfer_full_key_cost();
+    check_meta_tx_no_fn_call(&node, actions, tx_cost, amount, sender, relayer, new_account.clone());
+
+    // Check account exists
+    node.view_account(&new_account).expect("failed looking up account");
+}
+
+/// Try creating an implicit account with `CreateAction` which is not allowed in
+/// or outside meta transactions and must fail with `OnlyImplicitAccountCreationAllowed`.
+#[test]
+fn meta_tx_create_implicit_account_fails() {
+    let relayer = bob_account();
+    let sender = alice_account();
+    let new_account: AccountId = implicit_test_account();
+    let node = RuntimeNode::new(&relayer);
+
+    let actions = vec![Action::CreateAccount(CreateAccountAction {})];
+    let tx_result = node
+        .user()
+        .meta_tx(sender.clone(), new_account.clone(), relayer.clone(), actions.clone())
+        .unwrap();
+
+    let account_creation_result = &tx_result.receipts_outcome[1].outcome.status;
+    assert!(matches!(
+        account_creation_result,
+        near_primitives::views::ExecutionStatusView::Failure(TxExecutionError::ActionError(
+            ActionError { kind: ActionErrorKind::OnlyImplicitAccountCreationAllowed { .. }, .. }
+        )),
+    ));
+}
+
+/// Try creating an implicit account with a meta tx transfer and use the account
+/// in the same meta transaction.
+///
+/// This is expected to fail with `AccountDoesNotExist` as noted in NEP-366.
```

Comment:
What part of `NEP-366` is this, specifically?

My guess is that this is referring to https://github.com/near/NEPs/blob/550e428cfffafaa28bd84f2e882504a8679d01b6/neps/nep-0366.md?plain=1#L126-L130 but the wording in the NEP itself requires some reasoning about the text as a whole to understand this.

In particular, the expectation is that for DelegateAction to be valid, the `transaction.receiver_id` must match `delegate_action.sender_id`. My understanding then is that this then would prohibit creation of implicit accounts entirely (you cannot create a DelegateAction with an implicit account that does not yet exist?) However, shouldn’t the `tx.receiver_id == delegate_action.sender_id` check fail before anything relevant to account existence is checked?

How is it that the test below is possible at all, then?

Reply:
The link is correct, sadly it was not spelled out more clearly in the NEP itself.

> In particular, the expectation is that for DelegateAction to be valid, the transaction.receiver_id must match delegate_action.sender_id

Yes, in that case we would see `DelegateActionSenderDoesNotMatchTxReceiver` and we have a test for that case: https://github.com/near/nearcore/blob/master/runtime/runtime/src/actions.rs#L1365-L1408

>  My understanding then is that this then would prohibit creation of implicit accounts entirely (you cannot create a DelegateAction with an implicit account that does not yet exist?) 

Yes, the relayer has to create it first.

> How is it that the test below is possible at all, then?

In this test, we use the helper function `meta_tx` that accepts a meta tx sender, a relayer, and a  meta tx receiver. Sender and receiver naming here corresponds to the roles in the delegate action.
Inside `meta_tx` , we create the delegate action and wrap it in an outer transaction that uses the relayer as the sender of and the meta tx sender as the receiver. This ensures `tx.receiver_id == delegate_action.sender_id` is correct.

Reply:
I elaborated a bit more in the comment instead of a just useless reference to NEP-366 that doesn't really explain much.